rust/lance-encoding/src/encodings/physical/value.rs

```diff
@@ -67,11 +102,31 @@ impl PhysicalPageScheduler for ValuePageScheduler {
         let bytes = scheduler.submit_request(byte_ranges, top_level_row);
         let bytes_per_value = self.bytes_per_value;
 
+        let compressed = self.compressed;
+        let range_offsets = if compressed {
+            ranges
+                .iter()
+                .map(|range| {
+                    let start = (range.start as u64 * bytes_per_value) as usize;
+                    let end = (range.end as u64 * bytes_per_value) as usize;
+                    start..end
+                })
+                .collect::<Vec<_>>()
+        } else {
+            vec![]
+        };
+
         async move {
             let bytes = bytes.await?;
+            let data = if compressed {
+                ValuePageScheduler::decompress_with_ranges(&bytes, &range_offsets)?
```

Comment:
I am unsure if it is appropriate to handle decompression in the scheduler. However, since the compressed bytes read from disk may be reused multiple times, decompression should not occur in decode_into for optimal performance. Please advise on this. Thank you.

Reply:
Yeah, this is a bit tricky, I see your concern.  Unfortunately, we really shouldn't be doing the decompression here in the scheduler.  The scheduler is single threaded and by doing compression here we will fall behind in scheduling I/O.

Something simple, that should work, is to wrap the compression task in a Arc<Mutex<...>> that each of the decode tasks that need that range share.  Something roughly like...

```
struct ValuePageDecoder {
    bytes_per_value: u64,
    compressed_data: Vec<Bytes>,
    uncompressed_data: Arc<Mutex<Option<Bytes>>>,
}

impl ValuePageDecoder {

  // Called from decode_into
  fn get_uncompressed_bytes(&self) -> Result<Bytes> {
    let uncompressed_bytes = self.uncompressed_data.lock().unwrap();
    if uncompressed_bytes.is_none() {
      *uncompressed_bytes = self.decompress()?;
    }
    Ok(uncompressed_bytes.clone())
  }

}
```

This is not ideal because it means the decoder threads will be blocked for a bit but it should be good enough to start with.

A more complicated solution might be to have a dedicated thread pool for decompression.  Once I/O is available we start to decompress it in this thread pool.  I think we could do this as part of the I/O scheduler.  So we can have arguments to `submit_request` which explain that the data is compressed and includes the parameters needed to decompress it.  This would mean that compression is not just an "encoding" but something more fundamental but this is probably ok.  We probably also want to treat encryption similarly.  It also means we lose a bit of thread locality but I think that is inevitable.  Decompression / decryption is costly enough that we want to parallelize across columns.

Reply:
> Something simple, that should work, is to wrap the compression task in a Arc<Mutex<...>> that each of the decode tasks that need that range share

> This is not ideal because it means the decoder threads will be blocked for a bit but it should be good enough to start with.

Got it. I've considered this approach and its potential downsides. If it is deemed acceptable, I will revise the implementation accordingly.

Reply:
I made the change accordingly, and `uncompressed_bytes.clone()` seems to copy the data, so I use `Arc` to avoid the copy (but the lock is still hold during `decode_buffer`), could  you please check it out see if it fits? Thanks.